lib/Frontend/PrintingDiagnosticConsumer.cpp

```diff
@@ -61,24 +64,572 @@ namespace {
+      Out << "[insert '" << fixIt.getText() << "']";
+    } else if (fixIt.getText().empty()) {
+      Out << "[remove '" << SM.extractText(fixIt.getRange()) << "']";
+    } else {
+      Out << "[replace '" << SM.extractText(fixIt.getRange()) << "' with '"
+          << fixIt.getText() << "']";
+    }
+  }
+
+  /// Represents a single line of source code annotated with optional messages,
+  /// highlights, and fix-its.
+  class AnnotatedLine {
+    friend class AnnotatedFileExcerpt;
+
+    // A diagnostic message located at a specific byte in the line.
+    struct Message {
+      unsigned Byte;
+      DiagnosticKind Kind;
+      std::string Text;
+    };
+
+    // A half-open byte range which should be highlighted.
+    struct Highlight {
+      unsigned StartByte;
+      unsigned EndByte;
+    };
+
+    // A half-open byte range which should be replaced with the given text.
+    struct FixIt {
+      unsigned StartByte;
+      unsigned EndByte;
+      std::string Text;
+    };
+
+    unsigned LineNumber;
+    std::string LineText;
+    SmallVector<Message, 1> Messages;
+    SmallVector<Highlight, 1> Highlights;
+    SmallVector<FixIt, 1> FixIts;
+
+    // Adjust output color as needed if this byte is part of a fix-it deletion.
+    void applyStyleForLineByte(unsigned Byte, raw_ostream &Out, bool &Deleted) {
+      bool shouldDelete = false;
+
+      for (auto fixIt : FixIts) {
+        if (Byte >= fixIt.StartByte && Byte < fixIt.EndByte)
+          shouldDelete = true;
+      }
+
+      // Only modify deletions when we reach the start or end of
+      // a fix-it. This ensures that so long as the original
+      // SourceLocs pointed to the first byte of a grapheme cluster, we won't
+      // output an ANSI escape sequence in the middle of one.
+      if (shouldDelete != Deleted) {
+        Out.resetColor();
+        if (shouldDelete) {
+          Out.changeColor(ColoredStream::Colors::RED);
+        }
+      }
+      Deleted = shouldDelete;
+    }
+
+    // Insert fix-it replacement text at the appropriate point in the line.
+    bool maybePrintInsertionAfter(int Byte, raw_ostream &Out) {
+      for (auto fixIt : FixIts) {
+        if ((int)fixIt.EndByte - 1 == Byte) {
+          Out.changeColor(ColoredStream::Colors::GREEN, /*bold*/ true);
+          for (unsigned i = 0; i < fixIt.Text.size(); ++i) {
+            // Invert text colors for editor placeholders.
+            if (i + 1 < fixIt.Text.size() && fixIt.Text.substr(i, 2) == "<#") {
+              Out.changeColor(ColoredStream::Colors::GREEN, /*bold*/ true,
+                              /*background*/ true);
+              ++i;
+            } else if (i + 1 < fixIt.Text.size() &&
+                       fixIt.Text.substr(i, 2) == "#>") {
+              Out.changeColor(ColoredStream::Colors::GREEN, /*bold*/ true,
+                              /*background*/ false);
+              ++i;
+            } else {
+              Out << fixIt.Text[i];
```

Comment:
It may be helpful to have some other way of indicating that a certain span of text is a proposed fix-it insertion. At present, my first glance impression was that there was some syntax coloring going on, which doesn't immediately make it clear that the text isn't already in the source code.